corgidrp/spec.py

```diff
@@ -0,0 +1,410 @@
+    
+    (x_precis, y_precis) = (np.abs(xfwhm) / (2 * np.sqrt(2 * np.log(2))) / psf_peakpix_snr,
+                            np.abs(yfwhm) / (2 * np.sqrt(2 * np.log(2))) / psf_peakpix_snr)
+
+    return xfit, yfit, gauss2d_xfit, gauss2d_yfit, psf_peakpix_snr, x_precis, y_precis
+
+def compute_psf_centroid(dataset, initial_cent, output_path=None, verbose=False, halfwidth=10, halfheight=10):
+    """
+    Compute PSF centroids for a grid of PSFs and store them in a calibration file.
+
+    Args:
+        dataset (Dataset): Dataset containing 2D PSF images. Each image must include pri_hdr and ext_hdr.
+        initial_cent (dict): Dictionary with initial guesses for PSF centroids.
+                             Must include keys 'xcent' and 'ycent', each mapping to an array of shape (N,).
+                             Example:
+                                 initial_cent = {
+                                     'xcent': [12.0, 15.5, 10.3],
+                                     'ycent': [8.2, 14.1, 9.9]
+                                 }
+        output_path (str, optional): Path to write the resulting FITS file. Directory must exist.
+        verbose (bool): If True, prints fitted centroid values for each frame.
+        halfwidth (int): Half-width of the PSF fitting box. Default is 10.
+        halfheight (int): Half-height of the PSF fitting box. Default is 10.
+
+    Returns:
+        SpectroscopyCentroidPSF: Calibration object with fitted (x, y) centroids.
+    """
+    from corgidrp.data import Dataset
+
+    if not isinstance(dataset, Dataset):
+        raise TypeError("Input must be a corgidrp.data.Dataset object.")
+
+    xcent = np.asarray(initial_cent.get("xcent"))
+    ycent = np.asarray(initial_cent.get("ycent"))
+
+    if xcent is None or ycent is None:
+        raise ValueError("initial_cent dictionary must contain 'xcent' and 'ycent' arrays.")
+    if len(dataset) != len(xcent) or len(dataset) != len(ycent):
+        raise ValueError("Mismatch between dataset length and centroid guess arrays.")
+
+    centroids = np.zeros((len(dataset), 2))
+
+    # Use headers from the first frame in dataset
+    pri_hdr = dataset[0].pri_hdr.copy()
+    ext_hdr = dataset[0].ext_hdr.copy()
+
+    for idx, frame in enumerate(dataset):
+        psf_data = frame.data
+        xguess = xcent[idx]
+        yguess = ycent[idx]
+
+        xfit, yfit, *_ = fit_psf_centroid(
+            psf_data, psf_data,
+            xcent_template=xguess,
+            ycent_template=yguess,
+            xcent_guess=xguess,
+            ycent_guess=yguess,
+            halfwidth=halfwidth,
+            halfheight=halfheight
+        )
+
+        centroids[idx] = [xfit, yfit]
+
+        if verbose:
+            print(f"Slice {idx}: x = {xfit:.3f}, y = {yfit:.3f}")
+
+    calibration = SpectroscopyCentroidPSF(
+        centroids,
+        pri_hdr=pri_hdr,
+        ext_hdr=ext_hdr,
+        input_dataset=dataset
+    )
+
+    if output_path:
```

Comment:
we generally don't have save functions inside of step functions. There's a separate save function we use for all saving.